src/decode/stream.rs

```diff
@@ -0,0 +1,476 @@
+    fn flush(&mut self) -> std::result::Result<(), std::io::Error> {
+        if let Some(ref mut state) = self.state {
+            match state {
+                State::Init(_) => Ok(()),
+                State::Run(state) => state.decoder.output.get_mut().flush(),
+            }
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl std::convert::Into<std::io::Error> for Error {
+    fn into(self) -> std::io::Error {
+        std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", self))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    /// Test an empty stream
+    #[test]
+    fn test_stream_noop() {
+        let stream = Stream::new(Vec::new());
+        assert!(stream.get_ref().unwrap().is_empty());
+
+        let output = stream.finish().unwrap();
+        assert!(output.is_empty());
+    }
+
+    /// Test writing an empty slice
+    #[test]
+    fn test_stream_zero() {
+        let mut stream = Stream::new(Vec::new());
+
+        stream.write_all(&[]).unwrap();
+        stream.write_all(&[]).unwrap();
+
+        let output = stream.finish().unwrap();
+
+        assert!(output.is_empty());
+    }
+
+    /// Test processing only partial data
+    #[test]
+    fn test_stream_incomplete() {
+        let input = b"\x5d\x00\x00\x80\x00\xff\xff\xff\xff\xff\xff\xff\xff\x00\x83\xff\
+                         \xfb\xff\xff\xc0\x00\x00\x00";
+        // Process until this index is reached.
+        let mut end = 1;
+
+        // Test when we fail to provide the minimum number of bytes required to
+        // read the header. Header size is 13 bytes but we also read the first 5
+        // bytes of data.
+        while end < MAX_HEADER_LEN + START_BYTES {
+            let mut stream = Stream::new(Vec::new());
+            stream.write_all(&input[..end]).unwrap();
+            assert_eq!(stream.tmp_len, end);
+
+            let err = stream.finish().unwrap_err();
+            assert!(
+                err.to_string().contains("failed to read header"),
+                "error was: {}",
+                err
+            );
+
+            end += 1;
+        }
+
+        // Test when we fail to provide enough bytes to terminate the stream. A
+        // properly terminated stream will have a code value of 0.
+        while end < input.len() {
+            let mut stream = Stream::new(Vec::new());
+            stream.write_all(&input[..end]).unwrap();
+
+            // Header bytes will be buffered until there are enough to read
+            if end < MAX_HEADER_LEN + START_BYTES {
+                assert_eq!(stream.tmp_len, end);
+            }
+
+            let err = stream.finish().unwrap_err();
+            assert!(err.to_string().contains("failed to fill whole buffer"));
+
+            end += 1;
+        }
+    }
+
+    /// Test processing all chunk sizes
+    #[test]
+    fn test_stream_chunked() {
+        let small_input = b"Project Gutenberg's Alice's Adventures in Wonderland, by Lewis Carroll";
```

Comment:
This looks like a useful input to have as a test file, so that other unit tests can use it as well.

Reply:
Added a test file and include it with `include_bytes!("../../tests/files/small.txt")` given that it's a small input.